### ISSUE TITLE
`MapAt[f, expr, {i, j, ...}]` is broken when expr and i, j, etc are lists of lists. 

### ISSUE THREAD

**Issue by rocky** [Mathics3/mathics-core#1237]
## Description

The implementation of `MapAt[]` is horribly lacking. (I looked at who wrote this mess, and I see it was me!)

In particular, it does not handle lists of lists. 

## How to Reproduce

```
$ mathics -e 'MapAt[1&, {{0, 0}, {0, 0}}, {{1, 2}, {2, 1}}]'
{{0, 0}, {0, 0}}
```

## Expected behavior

The right answer for the previous section is `{ {0, 1}, {1, 0} }`. 

## Workarounds

An implementation could probably done in pure Mathic3 code.

## Priority

This is used in Combinatorica V0.9 which is how I discovered it. It appears that MapAt was added for the first time by me in trying to get Combinatorica working initially. 

## Additional context

Symja doesn't have the lameness and shallowness that Mathics3 currently has in its implementation. @axkr I would be grateful if you'd describe in human language, or at a high level, how SymJa implements MapAt. If possible, I'll try to follow that behavior in Mathics3's implementation.

**Comment by axkr** [comment]
> I would be grateful if you'd describe in human language, or at a high level, how SymJa implements MapAt.

I refactored the main algorithm here:
- https://github.com/axkr/symja_android_library/blob/master/symja_android_library/matheclipse-core/src/main/java/org/matheclipse/core/eval/util/positions/MapPositions.java

Maybe you can "reference" the part in a new Symja discussion entry, where we can discuss, what's going on in the Java code?

**Comment by rocky** [comment]
In short, I think I have found a solution or the beginning of a solution in https://github.com/Mathics3/mathics-core/pull/1239. 

I spent a day looking at MapAt and its wonky behavior.  The code has to be mindful of whether the third argument is an integer, a vector position, or a list of vector positions. The code has to tolerate changing the same location several times, and the shape of the replaced position can change dimension as well. 

And then a position might be out of range, but it also might specify a higher number of dimensions than the target expression has. 

I was interested in MapAt because it is used a bit (or at a lower-level core graph function) in Combinatorica v0.9. In Combinatorica v2.0.0 it is still around but not used so much.  And I don't see it used in Rubi and probably most other packages.
